app/routes/user/endorsements.server.ts

```diff
@@ -0,0 +1,292 @@
+/*!
+ * Copyright © 2022 United States Government as represented by the Administrator
+ * of the National Aeronautics and Space Administration. No copyright is claimed
+ * in the United States under Title 17, U.S. Code. All Other Rights Reserved.
+ *
+ * SPDX-License-Identifier: NASA-1.3
+ */
+
+import { tables } from '@architect/functions'
+import {
+  AdminListGroupsForUserCommand,
+  CognitoIdentityProviderClient,
+  ListUsersCommand,
+  ListUsersInGroupCommand,
+} from '@aws-sdk/client-cognito-identity-provider'
+import { getUser } from '../__auth/user.server'
+
+// models
+export type EndorsementRequest = {
+  endorserEmail: string
+  endorserSub: string
+  requestorEmail: string
+  requestorSub: string
+  created: number
+  status: EndorsementState
+}
+
+export type EndorsementState = 'approved' | 'pending' | 'rejected' | 'reported'
+
+const cognitoIdentityProviderClient = new CognitoIdentityProviderClient({})
+
+export class EndorsementsServer {
+  #sub: string
+  #currentUserEmail: string
+  #currentUserGroups: string[]
+
+  private constructor(sub: string, requestorEmail: string, groups: string[]) {
+    this.#sub = sub
+    this.#currentUserEmail = requestorEmail
+    this.#currentUserGroups = groups
+  }
+
+  static async create(request: Request) {
+    const user = await getUser(request)
+    if (!user)
+      throw new Response('not signed in', {
+        status: 403,
+      })
+    return new this(user.sub, user.email, user.groups)
+  }
+
+  /**
+   * Request a pending endorsement.
+   *
+   * Throws an HTTP error if:
+   *   - The current user already belongs to the GCN Circulars submitters group
+   *   - User `endorserSub` does not exist
+   *   - User `endorserSub` is not in the GCN Circulars submitters group
+   *   - There is an existing endorsement with the same requester and submitter sub but the state is not `pending`
+   */
+  async createEndorsementRequest(endorserSub: string) {
+    // Checking endorser sub:
+    if (!this.requestedEndorserIsValid(endorserSub))
+      throw new Response('Requested Endorser is not a verified submitter', {
+        status: 400,
+      })
+
+    const request = await this.#buildNewRequestFromEndorserSub(endorserSub)
+    const db = await tables()
+
+    const existingEndorsement = (await db.circular_endorsements.get({
+      requestorSub: this.#sub,
+      endorserSub: endorserSub,
+    })) as EndorsementRequest
+
+    if (existingEndorsement)
+      throw new Response(
+        'An endorsement already exists for the selected endorser',
+        {
+          status: 400,
+        }
+      )
```

Comment:
Only an error if the existing endorsement's status is not pending.

Check this using a ConditionExpression.